teuthology/suite/run.py

```diff
@@ -196,15 +196,58 @@ def choose_ceph_version(self, ceph_hash):
             log.info('skipping ceph package verification')
 
     def choose_teuthology_branch(self):
+        """Select teuthology branch, check if it is present in repo and
+        return the branch name value.
+
+        The branch name value is determined in the following order:
+
+        Use ``--teuthology-branch`` argument value if supplied.
+
+        Use ``TEUTH_BRANCH`` environment variable value if declared.
+
+        If yaml file ``qa/.teuthology`` can be found in suite repo
+        supplied with ``--suite-repo`` or ``--suite-dir`` and
+        has ``teuthology_branch`` record then use its value.
+
+        Use ``teuthology_branch`` value if it is set in the one
+        of the teuthology config files ``$HOME/teuthology.yaml``
+        or ``/etc/teuthology.yaml`` correspondingly.
+
+        Use ``--suite-branch`` argument value if supplied.
+
+        Use ``--ceph`` argument value.
+
+        Use ``master`` if the branch supplied with ``--suite-branch``
+        or ``--ceph`` is missing in the teuthology repo.
+
+        Generate exception if the branch is not present in the repo.
+        """
         teuthology_branch = self.args.teuthology_branch
-        if teuthology_branch and teuthology_branch != 'master':
-            if not util.git_branch_exists('teuthology', teuthology_branch):
-                exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
-                util.schedule_fail(message=str(exc), name=self.name)
-        elif not teuthology_branch:
+        if teuthology_branch:
+            if  teuthology_branch != 'master':
+                if not util.git_branch_exists('teuthology', teuthology_branch):
+                    exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
+                    util.schedule_fail(message=str(exc), name=self.name)
+        elif 'TEUTH_BRANCH' in os.environ:
+            teuthology_branch = os.environ.get('TEUTH_BRANCH')
+        else:
             # Decide what branch of teuthology to use
-            if util.git_branch_exists('teuthology', self.args.ceph_branch):
-                teuthology_branch = self.args.ceph_branch
+            check_branch = None
+            branch_config = self.suite_repo_path + '/qa/.teuthology'
```

Comment:
to avoid the impression that this file could contain other options, perhaps it could be named something more specific, like /qa/.teuthology_branch

Reply:
... and changed to contain just the branch name instead of "teuthology_branch: foo" key/value pair.

Reply:
that makes sense

Reply:
pr updated

Reply:
but @jdurgin @idryomov nevertheless I changed it, I like more to have `.teuthology` file, this is a standard practice for modern tools to put repo configuration, like git, jenkins, travis, etc. And I need more configuration options rather than teuthology_branch in it.

Reply:
Just to be clear, I didn't object to having a *real* override file.  What I picked on was the fact that `.teuthology` looked like an override file but only handled one option.

At a high level, there are two ways to do this: either something really specific to `teuthology_branch` option or a general solution that would handle all options.  Your original approach was kind of in the middle: it had the appearance of a general solution but couldn't handle other options.